### PR TITLE
Pull in working* model

### PR DESCRIPTION
**Pull request by willschwarzer**

### CODE REVIEW

File: experiments/vectorized/concept.py [hunk 1]
```diff
@@ -213,15 +227,20 @@ def compute_self_att_loss(alphas):
     def train(epoch=-1, backprop=True):
         """ Train model for a single epoch.
         """
+        ### Model is declared in global space
```
Comment (jayelm) [inline experiments/vectorized/concept.py:230]: Yup, this is a [function closure](https://en.wikipedia.org/wiki/Closure_(computer_programming))

File: experiments/vectorized/concept_teacher.py [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+        # Get the batch size (and make sure it's the same for all data)
+        batch_size = stims.shape[0]
+        assert(batch_size == language.shape[0] == lang_lengths.shape[0] == labels.shape[0])
+        # "Unfold" the sequence so we have a 2d matrix
+        logits_2d = logits.view(batch_size * (max_seq_len - 1), -1)
+        ### TODO we probably don't need to convert language to longs here
+        ### (it should already be longs, since we never converted to floats
+        language_1d = language.long().view(batch_size * (max_seq_len - 1))
+
+        # Cross entrops is your loss function - in short, you pay a penalty if you put probability amss onl #TODO ?
```
Comment (jayelm) [inline experiments/vectorized/concept_teacher.py:297]: In classification minimizing the cross entropy is the same as maximizing the log probability of the prediction of the correct class for each example

What the PyTorch implementation of cross entropy does is apply a log softmax on the unnormalized logits first, then uses negative log likelihood loss

Comment (jayelm) [reply experiments/vectorized/concept_teacher.py:297]: https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L1989-L1991

File: experiments/vectorized/concept_teacher.py [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+            x_l_lengths = None
+        ### get the (batch_size) tensor (basically list) of stimuli
+        stims = construct_stim_reps(batch)
+        labels = batch.__dict__['labels'].float()
+        if args.cuda:
+            stims = stims.to('cuda')
+            language = language.to('cuda')
+            lang_lengths = lang_lengths.to('cuda')
+            labels = labels.to('cuda')
+        breakpoint()
```
Comment (jayelm) [inline experiments/vectorized/concept_teacher.py]: ?

File: experiments/vectorized/concept_teacher.py [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+                'stim_model_type': 'featureMLP',
+                'kwargs': kwargs
+            },
+            is_best,
+            os.path.join(args.out_dir, kwargs['model_id']), 
+            'checkpoint.pth.tar',
+            'model_best.pth.tar'
+        )
+    '''
+    ### TODO why are so many words mapping to the unknown character?
```
Comment (jayelm) [inline experiments/vectorized/concept_teacher.py]: Is there some filtering such that we only keep words that occur more than a few times in the training set (so a `min_count` parameter or something?)

File: experiments/vectorized/concept_teacher.py [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+        ### get the (batch_size) tensor (basically list) of stimuli
+        stims = construct_stim_reps(batch)
+        labels = batch.__dict__['labels'].float()
+        if args.cuda:
+            stims = stims.to('cuda')
+            language = language.to('cuda')
+            lang_lengths = lang_lengths.to('cuda')
+            labels = labels.to('cuda')
+        breakpoint()
+        orig_lang = fields['text'][1].reverse(language)
```
Comment (jayelm) [inline experiments/vectorized/concept_teacher.py:344]: Why do we access the first element of `fields['text']` both here and elsewhere?

File: models/student/lfl/teacher.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        # self.mlp = MLP(
+        #     i_dim=kwargs['o_dim_l'] + kwargs['o_dim_s'],
+        #     h_dim=kwargs['hidden_dim_student'],
+        #     o_dim=kwargs['output_dim'],
+        #     d_prob=kwargs['d_prob_student'],
+        #     batch_norm=kwargs['batch_norm'],
+        #     num_layers=kwargs['num_layers_student'],
+
+    ### forward: defines how the module transforms input to output
+    ### technically just a function on any arbitrary input, can give any arbitrary output?
```
Comment (jayelm) [inline models/student/lfl/teacher.py]: yup

you also don't necessarily need useful model functionality in the `forward` method (e.g. you could implement a method like `set_loss`) - `model.forward(inputs)` is sugar for `model(inputs)` and also provides some pre/post-call hooks which are useful for debugging and analysis

File: models/teacher/language/rnn_decoder.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+        # PyTorch RNNs return two values, you can look up the details here:
+        # https://stackoverflow.com/questions/48302810/whats-the-difference-between-hidden-and-output-in-pytorch-lstm
+        ### (The details are that the first value is the output, i.e. a tensor of output values for each time step t
+        ### the second value is just the hidden state at the very end, in case you want to feed it into another rnn)
+
+        # The two inputs are (1) your packed input sequence; (2) the hidden
+        # state initialization (if any). If the 2nd argument is left out we
+        # default to 0s. (Show and tell does not initialize the hidden state;
+        # this code does).
+        ### TODO how does show and tell do it without initializing the hidden state?
```
Comment (jayelm) [inline models/teacher/language/rnn_decoder.py]: Ah, yes, there are variants of the model where you don't initialize the hidden state, rather you just provide the image representation as the first input token. In that case you most likely initialize the hidden state to all 0s

File: models/teacher/language/rnn_decoder.py [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+        ### (second element of return tuple is a list of sequence lengths,
+        ### which we already have)
+        output = rnn_utils.pad_packed_sequence(packed_output, batch_first = True)
+        ### contiguous() places all values next to each other in memory, rather than
+        ### potentially having references instead of the actual values
+        output = output[0].contiguous()
+        ### shape: (batch_size, max seq length, hidden dim)
+
+        # Sort the language back in the order we saw earlier
+        if batch_size > 1:
```
Comment (jayelm) [inline models/teacher/language/rnn_decoder.py]: If you removed `batch_size > 1` up top, should you remove it here too?

File: models/teacher/language/rnn_decoder.py [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+
+            # (B,L,D) to (L,B,D)
+            #inputs = inputs.transpose(0, 1)
+
+            # compute embeddings
+            inputs = self.embedding(inputs)
+
+            # We'll sample until we observe an end of sentence token or we hit
+            # MAX_SEQ_LEN, whichever is faster. After MAX_SEQ_LEN we give up.
+            for i in range(MAX_SEQ_LEN):
```
Comment (jayelm) [inline models/teacher/language/rnn_decoder.py:179]: Is `MAX_SEQ_LEN = 20` limiting the decoder's output at all? I.e. do we run into `MAX_SEQ_LEN` before the decoder finishes generating sentences?

20 was picked for a different task - might be worth increasing this. The safest thing to do is to just set `MAX_SEQ_LEN` to be the length of the longest sequence in your training data.

File: utils/dataloaders/vectorized/concept_load_dataset.py [hunk 10]
```diff
@@ -78,17 +78,26 @@ def construct_field(
         raise Exception('Invalid Field Type')
 
 
+### Note: we're using torchtext for its torchtext.data ("data") library
+### Apparently torchtext is a "text preprocessing" library, designed to work with any DL library
+### http://anie.me/On-Torchtext/ <----- excellent tutorial
```
Comment (jayelm) [inline utils/dataloaders/vectorized/concept_load_dataset.py:83]: Allen's a member of the lab, FYI!
